README.md
# mapnik-omnivore
Node module that returns metadata of spatial files.
Version format follows [Semantic Version](http://semver.org/)

[![Build Status](https://travis-ci.org/mapbox/mapnik-omnivore.svg?branch=master)](https://travis-ci.org/mapbox/mapnik-omnivore)
[![Coverage Status](https://coveralls.io/repos/mapbox/mapnik-omnivore/badge.svg?branch=master)](https://coveralls.io/r/mapbox/mapnik-omnivore?branch=master)

Currently supports the following file types:
- `geojson`
- `topojson`
- `kml`
- `gpx`
- `tif`
- `vrt` : raster files listed in the VRT file must be in the same directory as the VRT file
- `csv` : must be valid geo CSV, and can be in the form of `csv`, `txt`, or `tsv`
- `shp` : In order to set the projection, the `.prj` file must be in the same directory and have the same name as the `.shp` file


## Javascript Usage

```
var mapnikOmnivore = require('@mapbox/mapnik-omnivore'),
    path = require('path');

var file = path.resolve('test/data/zip/world_merc/world_merc.shp');

mapnikOmnivore.digest(file, function(err, metadata){
	if (err) return callback(err);
	else {
		console.log('Metadata returned!');
		console.log(metadata);
	}
});
```

### mapnikOmnivore.digest(filepath, callback)
- filepath `required`
- callback `(err, metadata)`


#### Example of returned `metadata`
```
{ filesize: 428328,   // size of file in bytes
  projection: '+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over',
  filename: 'world_merc',
  center: [ 0, 12.048603815490733 ],
  extent: [ -180, -59.47306100000001, 180, 83.57026863098147 ],
  json: { 
  	vector_layers: [ { 
      id: 'world_merc',
      description: '',
      minzoom: 0,
      maxzoom: 22,
      fields: { 
        FIPS: 'String',
        ISO2: 'String',
        ISO3: 'String',
        UN: 'Number',
        NAME: 'String',
        AREA: 'Number',
        POP2005: 'Number',
        REGION: 'Number',
        SUBREGION: 'Number',
        LON: 'Number',
        LAT: 'Number' } 
    } ] 
  },
  minzoom: 0,    // calculates the optimal minimum and
  maxzoom: 5,   // maximum zoom levels for the file
  layers: [ 'world_merc' ],
  dstype: 'shape',
  filetype: '.shp' }
```

Raster files will include a `raster` object:
```
{ filesize: 1494,
  projection: '+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=23 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs',
  raster: 
   { pixelSize: [ 7.502071930146189, 7.502071930145942 ],
     bandCount: 1,
     bands: [ [Object] ],
     nodata: null,
     origin: [ -1134675.2952829634, 2485710.4658232867 ],
     width: 984,
     height: 804,
     units: { linear: [Object], angular: [Object] } },
  filename: 'sample',
  center: [ -110.32476292309875, 44.56502238336985 ],
  extent: 
   [ -110.3650933429331,
     44.53327824851143,
     -110.28443250326441,
     44.596766518228264 ],
  minzoom: 0,
  maxzoom: 13,
  dstype: 'gdal',
  filetype: '.vrt',
  layers: [ 'sample' ] }

```

## CLI Usage
```
$ npm install --global @mapbox/mapnik-omnivore
$ digest <filepath>
# Prints a JSON string
```

## Generating Mapnik XML

1) Install the https://github.com/mapbox/tilelive-omnivore

2) Use the command line tool it provides, called `mapnik-omnivore` to generate Mapnik XML


## Install
```
npm install @mapbox/mapnik-omnivore
```

## Tests
`npm test`


CHANGELOG.md
# Changelog

## 10.0.0

- Limit maxzoom output to z16 for any vector source
- Remove node8 support, test with node12

## 9.1.1

- Update mapnik install to allow 4.x or 3.x

## 9.1.0

- Updated dependencies to allow install with node v10
- Add linter
- Remove test for node 4 + 6, add tests for node 8 + 10

## 9.0.1

- try/catch around more reprojection errors

## 9.0.0

- Updated to mapnik 3.7.0
- Dropped support for Windows binaries

## 8.5.0

- Clamp minzoom to z0 for all raster data
- Try/catch forward projections in raster extent method

## 8.4.2

- Clamp minzoom to z0 for vector data files under 10 MB

## 8.4.1

- Increase raster maxzoom to z25

## 8.4.0

- Upgraded to use mapnik 3.6.0

## 8.3.0

- Shorten the CSV errors when Mapnik cannot find proper column headers [#157](https://github.com/mapbox/mapnik-omnivore/issues/157)
- Introduce an internal framework for determining feature-specific zoom levels, just starting with small point datasets for now [#151](https://github.com/mapbox/mapnik-omnivore/issues/151), [#140](https://github.com/mapbox/mapnik-omnivore/issues/140)
- Update to @mapbox/mapbox-file-sniff and @mapbox/sphericalmercator packages
- Update appveyor to Node v4 and v6, no longer testing on Node v0.10.x

## 8.2.2

- Update to collect all feature's properties from GeoJSON files

## 8.2.1

- Increase raster maxzoom to z23

## 8.2.0

- Move module to @mapbox namespace
- Fixes incorrect ft + mi + km coordinate conversion for raster pixel sizes

## 8.1.0

 - Update to node v6; mapnik 3.5.14; gdal 0.9.3; and mapbox-file-sniff 0.5.2

## 8.0.0

 - Will no longer provide a recommended max zoom level less than 6

## 7.4.0

 - Adjust maxzoom limits for rasters. New fillzoom max: 21, New maxzoom: 22

## 7.3.0

 - Updated to use 3.5.0 mapnik

## 7.1.1

 - Upgraded to node-srs@1.1.0 to pull in node-gdal@0.8.0

## 7.1.0

 - Upgraded to node-gdal@0.8.0 and queue-async@1.1.1

## 7.0.0

 - Adjustments to maxzoom calculations for rasters that result in less oversampling and may yield different ranges than previous versions
 - Fixes a bug in shapefile analysis when mapnik cannot parse the file's projection information

## 6.3.2

 - Now handles more types of GeoJSON parsing errors (both datasource creation and when calling `ds.describe()`) (#128)
 - Upgraded to node-srs@1.x
 - Compatibility with latest node-mapnk@3.4.10 / Mapnik v3.0.9

## 2.2.1

 - Updated node-mapnik to 3.1.1
 - Updated node-gdal to 0.3.1
 - Updated node-srs to 0.4.5


lib/datasourceProcessor.js
var fs = require('fs'),
    path = require('path'),
    sphericalMerc = new(require('sphericalmercator')),
    invalid = require('./invalid'),
    srs = require('srs'),
    gdal = require('gdal'),
    mapnik = require('mapnik');

// Register datasource plugins
mapnik.register_default_input_plugins();

/**
 * Initializes the module
 * @param file (filepath)
 * @param filesize
 * @param filetype
 * @returns metadata {filesize, projection, filename, center, extent, json, minzoom, maxzoom, layers, dstype, filetype}
 */
function init(file, filesize, filetype, callback) {
    getDatasourceConfigs(file, filesize, filetype, function(err, configs) {
        if (err) return callback(err);
        return callback(null, configs);
    });
};
/**
 * Obtains metadata depending on file type
 * @param file (filepath)
 * @param filesize
 * @param filetype
 * @returns metadata {filesize, projection, filename, center, extent, json, minzoom, maxzoom, layers, dstype, filetype}
 */
function getDatasourceConfigs(file, filesize, filetype, callback) {
    getProjection(file, filetype, function(err, proj) {
        if(err) return callback(err);
        var name = path.basename(file, path.extname(file));
        if(filetype === '.geojson') name = name.replace('.geo', '');
                
        if(filetype === '.tif' || filetype === '.vrt') {
            processRasterDatasource(file, filesize, name, proj, filetype, function(err, info) {
                if(err) return callback(err);
                return callback(null, {
                    filesize: filesize,
                    projection: info.proj,
                    raster: info.raster,
                    filename: name,
                    center: info.center,
                    extent: info.extent,
                    minzoom: info.minzoom,
                    maxzoom: info.maxzoom,
                    dstype: info.dstype,
                    filetype: filetype,
                    layers: [name]
                });
            }); 
        // shapefile datasource
        } else if (filetype === '.shp' || filetype === '.csv') {
            processDatasource(file, filesize, name, proj, filetype, function(err, info) {
                if (err) return callback(err);
                return callback(null, {
                    filesize: filesize,
                    projection: proj,
                    filename: name,
                    center: info.center,
                    extent: info.extent,
                    json: info.json,
                    minzoom: info.minzoom,
                    maxzoom: info.maxzoom,
                    layers: info.layers,
                    dstype: info.dstype,
                    filetype: filetype
                });
            });
        // else OGR datasource
        } else {
            processOgrDatasource(file, filesize, proj, filetype, function(err, info) {
                if (err) return callback(err);
                return callback(null, {
                    filesize: filesize,
                    projection: proj,
                    filename: name,
                    center: info.center,
                    extent: info.extent,
                    json: info.json,
                    minzoom: info.minzoom,
                    maxzoom: info.maxzoom,
                    layers: info.layers,
                    dstype: info.dstype,
                    filetype: filetype
                });
            });
        }
    });
};
/**
 * Obtains projection based on filetype
 * @param file (filepath)
 * @param filetype
 * @returns projection String
 */
function getProjection(file, filetype, callback) {
    if (filetype === '.geojson' || filetype === '.csv') return callback(null, '+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs');
    //else if shape
    else if (filetype === '.shp') {
        projectionFromShape(file, function(err, proj) {
            if (err) return callback(err);
            return callback(null, proj);
        });
    //else if raster
    }
    else if(filetype === '.tif' || filetype === '.vrt') {
        projectionFromRaster(file, function(err, proj) {
            if (err) return callback(err);
            return callback(null, proj);
        });
    //else kml and gpx
    }
    else return callback(null, '+init=epsg:4326');
};
/**
 * Obtains extent from the datasource and the center point (lat/lng)
 * @param ds (Mapnik datasource)
 * @param projection
 * @param filetype
 * @returns results (results object that contains extent and center point)
 */
function getCenterAndExtent(ds, projection) {
    var extent;
    // Convert datasource extent to lon/lat when saving
    var fromProj = new mapnik.Projection(projection);
    var toProj = new mapnik.Projection('+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs');
        
    if (fromProj === toProj) {
        try {
            extent = ds.extent();
        } catch (err) {
            return invalid('Error obtaining extent of Mapnik datasource.');
        }
    } else {
        try {
            var trans = new mapnik.ProjTransform(fromProj, toProj);
            //Bounding box
            extent = trans.forward(ds.extent());
        } catch (err) {
            return invalid('Error obtaining extent of Mapnik datasource.');
        }
    }
    
    //Center point of bounding box (extent)
    var center = [0.5 * (extent[0] + extent[2]), 0.5 * (extent[1] + extent[3])];
    var results = {
        extent: extent,
        center: center
    }
    return results;
};
/**
 * Gets min/max zoom levels based on filesize
 * @param bytes (size of file)
 * @param extent
 * @returns callback(err, minzoom, maxzoom)
 */
function getMinMaxZoom(bytes, extent, callback) {
    //Threshold for largest avg tilesize allowed
    var maxSize = 500 * 1024;
    var minzoom = 0;
    var maxzoom = 22;
    //Calculate min/max zoom
    for (i = maxzoom; i >= 0; i--) {
        var currentzoom = i;
        //Defaulting srs to WGS84 based on how the extent is set in mapnikSetup.getCenterAndExtent() function
        var bounds = sphericalMerc.xyz(extent, currentzoom, false, 4326);
        var x = (bounds.maxX - bounds.minX) + 1;
        var y = (bounds.maxY - bounds.minY) + 1;

        //Number of tiles within current zoom level
        var tiles = x * y;

        if (tiles <= 0) {
            return callback(invalid('Error calculating min/max zoom: Bounds invalid'));
        }
        if (bytes <= 0) {
            return callback(invalid('Error calculating min/max zoom: Total bytes less than or equal to zero'));
        }
        //Average tilesize within current zoom level
        var avgTileSize = bytes / tiles;

        //The idea is that tilesize of ~1000 bytes is usually the most detail needed, and no need to process tiles with higher zoom
        if (avgTileSize < 1000) {
            maxzoom = currentzoom;
        }
        //If avg tile size is small enough to get to one remaining tile, just set minzoom to zero.
        if (avgTileSize > maxSize) {
            minzoom = currentzoom;
            return callback(null, minzoom, maxzoom);
        } else if (tiles === 1 || i === 0) {
            minzoom = 0;
            return callback(null, minzoom, maxzoom);
        }
    };
};
/**********************
 * Shape and CSV files
 **********************/
 /**
 * Creates a mapnik Datasource and uses that to obtain the file's spatial properties
 * @param file (filepath)
 * @param filesize
 * @param name
 * @param proj
 * @param filetype
 * @returns object that contains spatial properties of the file {extent, center, json, minzoom, maxzoom, layers, dstype}
 */
function processDatasource(file, filesize, name, proj, filetype, callback) {
    var options = {};
    var dstype;
    //setup shape datasource options
    if(filetype === '.shp'){
        dstype = 'shape';
        options = {
            type: dstype,
            file: file,
            layer: name
        };
    //else setup csv datasource options
    } else {
        dstype = 'csv';
        options = {
            type: dstype,
            file: file,
            filesize_max:10,
            layer: name
        };
    }
    var ds = new mapnik.Datasource(options);
    //Center of bounding box/extent
    var results = getCenterAndExtent(ds, proj, filetype);
    //Get fields and setup vector layers json
    var actual = ds.describe();
    var fields = actual.fields;

    getMinMaxZoom(filesize, results.extent, function(err, min, max) {
        if (err) return callback(err);
        var json = {
            "vector_layers": [{
                "id": name,
                "description": "",
                "minzoom": 0,
                "maxzoom": 22,
                "fields": fields,
                "center": results.center,
                "extent":results.extent
            }]
        };
        //Values to input into xml for Mapnik
        return callback(null, {
            extent: results.extent,
            center: results.center,
            json: json,
            minzoom: min,
            maxzoom: max,
            layers: [options.layer],
            dstype: dstype
        });
    });
};
 /**
 * Obtains projection from a shapefile by reading the .prj file in the same directory
 * @param file (filepath)
 * @returns result.proj4
 */
function projectionFromShape(file, callback) {
    var fileDir = path.dirname(file);
    //Assumes the .prj file has the same name as the .shp file
    var projFile = fileDir + "/" + (path.basename(file, path.extname(file))) + '.prj';
    fs.readFile(projFile, function(err, data) {
        if(err) return callback(invalid('Missing projection file.'));
        var result;
        try {
            result = srs.parse(data);
        } catch (err) {
            return callback(invalid('Invalid projection file.'));
        }
        //TODO: check if the current projection is something we can deal with
        //...what are some projections we can't deal with?
        //Currently handles issues with ESRI projections
        if (result.proj4 === undefined) {
            var esriAppended = "ESRI::" + data.toString();
            projectionFromShape(esriAppended);
        } else {
            //Yay, Valid!
            return callback(null, result.proj4);
        }
    });
};


/********************
 * Raster/gdal files
 ********************/
/**
 * Obtains the Raster/GDAL file's spatial properties
 * @param file (filepath)
 * @param filesize
 * @param name
 * @param info Object
 * @param filetype
 * @returns object that contains spatial properties of the file {extent, center, raster properties, projection, minzoom, maxzoom, dstype}
*/
function processRasterDatasource(file, filesize, name, info, filetype, callback) {
    var options = {
        type: 'gdal', 
        file: file
    };
    try{
        var ds = new mapnik.Datasource(options);
        var ref = gdal.SpatialReference.fromProj4(info.proj4);
        var results = getCenterAndExtent(ds, info.proj4, filetype);
    } catch(err){
        return callback(invalid(err));
    }
    //calculate source's min/max zoom
    getMinMaxZoomGDAL(info.raster.pixelSize, results.center, info.proj4, function(err, min, max) {
        if (err) return callback(err);
        //Values to input into xml for Mapnik
        return callback(null, {
            extent: results.extent,
            center: results.center,
            raster: info.raster,
            proj: info.proj4,
            minzoom: min,
            maxzoom: max,
            dstype: 'gdal'
        });
    });
};
/**
 * Calculates a GDAL source's min and max zoom level using native pixel size and converting to google mercator (meters)
 * @param pixelSize (source's pixel size Array)
 * @param center (source's center point Array)
 * @param proj (source's srs String)
 * @returns callback(err, minzoom, maxzoom)
 */
function getMinMaxZoomGDAL(pixelSize, center, proj, callback){
    var refFrom = new gdal.SpatialReference.fromProj4(proj);
    var refTo = new gdal.SpatialReference.fromEPSG(3857);
    var transform = new gdal.CoordinateTransformation(refFrom, refTo);
    
    // Grab half of pixelsize and add/subtract to center longitude to create a horizontal line 
    var halfPixelsPerUnit = pixelSize[0]/2;
    var line = new gdal.LineString();
    line.points.add(center[0], (center[1] - halfPixelsPerUnit));
    line.points.add(center[0], (center[1] + halfPixelsPerUnit));
    // Transform native res pixel distance to google mercator pixel distance in meters
    line.transform(transform);
    var mercatorPixelSize = line.getLength();

    // After transforming pixelsize, earth circumference will always be in meters...?
    var circ = 40075000;
    var lat = center[1];
    var res;
    var maxzoom;

    //iterate through zoom levels to find threshold and set maxzoom
    for(var zoom = 19; zoom >=0; zoom--){
        // calculate resolution (meters/pixel) for each zoom level
        // S=C*cos(y)/2^(z+8)...from http://wiki.openstreetmap.org/wiki/Zoom_levels
        // calculates for lat north AND south of equator
        res = circ * Math.cos(lat*Math.PI/180) / Math.pow(2,(zoom+8));
        //use the source's resolution as the threshold
        if(res >= mercatorPixelSize){
            maxzoom = zoom+1;
            return callback(null, Math.max(0, maxzoom-6), maxzoom);
        }
        if (zoom <= 0) return callback(invalid('Failed to set min/max zoom'));
    }
};
/**
 * Detects unit type of the gdal source, using the srs/projection
 * @param srs
 * @returns unit type (String) 
 */
function getUnitType(srs){
    var possibleUnits = ['m','ft','mi','km','us-ft','us-mi'];
    for(var i = 0; i < possibleUnits.length; i++){
        if(srs.indexOf("+units=" + possibleUnits[i]) !== -1) return possibleUnits[i];
    }
    if(srs.indexOf("+units=") === -1 && srs.indexOf("+proj=longlat") !== -1) return 'decimal degrees';
    //Default to meters for now, if nothing matches
    else return 'm';
};
/**
 * Obtains projection from a raster by using node-gdal lib
 * @param file (filepath)
 * @returns result.proj4 and raster properties object
 */
function projectionFromRaster(file, callback) {
    var ds;
    var proj;
    var band;
    var bandCount;
    var bandStats;
    var width;
    var height;
    var geotransform;
    var bands;
    var nodata;
    var unitType;
    var metadata;
    var fileList;
    var srsResult;
    
    try {
        ds = gdal.open(file);
        geotransform = ds.geoTransform;
        bandCount = ds.bands.count();
        width = ds.rasterSize.x;
        height = ds.rasterSize.y;
        proj = ds.srs.toWKT();
        fileList = ds.getFileList();
        //get srs string
        srsResult = srs.parse(proj);

        //get native units/pixel
        var ref = gdal.SpatialReference.fromProj4(srsResult.proj4);
        var linearUnits = ref.getLinearUnits();
        var angularUnits = ref.getAngularUnits();
        
        //get bands and bandStats
        bands = getBands(ds, bandCount);
    } catch(err) {
        return callback(invalid('Invalid gdal source. ' + err.message));
    }   
    if (bands instanceof Error) return callback(bands);

    //TODO: check if the current projection is something we can deal with
    //...what are some projections we can't deal with?
    if(srsResult.proj4 === undefined){
        return callback(invalid('Undefined projection from gdal source.'));
    } else return callback(null, {
        'proj4': srsResult.proj4,
        'raster':{
            'pixelSize': [geotransform[1],-geotransform[5]], 
            'bandCount': bandCount,
            'bands': bands,
            'nodata': bands[0].nodata,
            'origin': [geotransform[0], geotransform[3]],
            'width': width,
            'height': height,
            'units':{
                'linear': linearUnits,
                'angular': angularUnits
            }
        }
    });
};
/**
 * Iterates through source's bands and obtains band properties
 * @param ds (Datasource)
 * @param bandcount
 * @returns bands Array
 */
function getBands(ds, bandCount){
    var bands = [];
    var band;
    var unitType;
    var bandStats;
    var nodata;
    for(var i = 1; i <= bandCount; i++){
        band = ds.bands.get(i);
        nodata = band.noDataValue;
        unitType = ds.bands.get(i).unitType;
        //Error check file sources
        try { 
            bandStats = band.getStatistics(false, true);
        } catch(err){
            return invalid("Error getting statistics of band. 1 or more of the VRT file's relative sources may be missing: " + err.message);
        };
        //add to bands array
        bands.push({
            //commenting out band for now, seems to be erroring out when trying to stringify C++ object
            //'band':band, 
            'stats':bandStats,
            'scale':band.scale,
            'unitType':unitType,
            'rasterDatatype':band.dataType,
            'categoryNames':band.categoryNames,
            'hasArbitraryOverviews':band.hasArbitraryOverviews,
            'overviews':band.overviews,
            'nodata':nodata,
            'id':band.id,
            'blockSize':band.blockSize,
            'color':band.colorInterpretation
        });
    };
    return bands;
};

/************
 * OGR files
 ************/
/**
 * Obtains the OGR file's spatial properties
 * @param file (filepath)
 * @param filesize
 * @param name
 * @param proj
 * @param filetype
 * @returns object that contains spatial properties of the file {extent, center, json, minzoom, maxzoom, layers, dstype}
 */
function processOgrDatasource(file, filesize, proj, filetype, callback) {
    getDatasourceProperties(file, filetype, function(err, layers, ds) {
        if (err) return callback(err);
        var results = getCenterAndExtent(ds, proj, filetype);
        //if error returned from synchronous function, send Error to callback
        if (results instanceof Error) {
            return callback(results);
        };
        getMinMaxZoom(filesize, results.extent, function(err, min, max) {
            if (err) return callback(err);
            //object to hold layer attributes
            var json = { "vector_layers": [] };
            var actual = ds.describe();
            var fields = actual.fields;
            layers.forEach(function(layer) {
                //setup vector layers json
                json.vector_layers.push({
                    "id": layer.split(' ').join('_'),
                    "description": "",
                    "minzoom": 0,
                    "maxzoom": 22,
                    "fields": fields,
                    "center": results.center,
                    "extent":results.extent
                });
            });
            return callback(null, {
                extent: results.extent,
                center: results.center,
                json: json,
                minzoom: min,
                maxzoom: max,
                layers: layers,
                dstype: 'ogr'
            });
        });
    });
};
 /**
 * Creates a mapnik Datasource and uses that to obtain the file's spatial properties
 * @param file (filepath)
 * @param name
 * @param filetype
 * @returns layers (file's layer names) 
 *          json (vector_layers array) 
 *          ds (mapnik Datasource)
 */
function getDatasourceProperties(file, filetype, callback) {
    var dstype = 'ogr';
    //Get layers for kml or geojson/topojson
    if (filetype === '.kml' || filetype === '.geojson') {
        var options = {
            type: dstype,
            file: file
        };
        //Get KML/geojson layer names from the OGR error message...for now
        getLayers(options, function(err, layers) {
            if (err) return callback(err);
            getDatasource(options, layers, function(err, ds) {
                if (err) return callback(err);
                else return callback(null, layers, ds);
            });
        });
        //Get layers for .gpx
    } else if (filetype === '.gpx') {
        var options = {
            type: dstype,
            file: file
        };
        //GPX files can only have these layers
        var layers = ['waypoints', 'routes', 'tracks', 'route_points', 'track_points'];
        getDatasource(options, layers, function(err, ds) {
            if (err) return callback(err);
            return callback(null, layers, ds);
        });
    }
};

//Sole purpose is to obtain layer names from OGR error message
function getLayers(options, callback) {
    var layers_array;
    var error;
    //Expect an error in order to obtain layer names from the OGR Error string...for now
    try {
        var ds = new mapnik.Datasource(options);
    } catch (err) {
        if (err.message && err.message.indexOf('OGR Plugin: missing <layer>') != -1) {
            var layers = err.message.split("are: ")[1];
            //trim whitespaces before each layer
            layers = layers.replace(/\s*,\s*/g, ',');
            //remove quotes
            var layer_names = layers.split('\'').join('');
            //designate each index of the array by splitting by commas
            layers_array = layer_names.split(',');
        } else error = err;
    }
    if (error === undefined) return callback(null, layers_array);
    else return callback(invalid('Error obtaining layer names'));
};

//Create a datasource, which is later used to obtain extent and center of the source
function getDatasource(options, layers, callback) {
    var validDatasource;
    var error;
    layers.forEach(function(layer) {
        options.layer = layer;
        try {
            var ds = new mapnik.Datasource(options);
            //Check to see that layer actually exists
            var features = [];
            var featureset = ds.featureset();
            var feature = featureset.next();
            if(feature === undefined) {
                var index = layers.indexOf(layer);
                if (index > -1) {
                    layers.splice(index, 1);
                }
            } else validDatasource = ds;
        } catch (err) {
            error = invalid('Error creating Mapnik Datasource: ' + err.message);
        }
    });

    if (error === undefined && validDatasource !== undefined) return callback(null, validDatasource);
    else if (validDatasource === undefined) return callback(invalid('Source appears to have no features data.'));
    else return callback(error);
};
module.exports = {
    init: init,
    getCenterAndExtent: getCenterAndExtent,
    getDatasource: getDatasource,
    getMinMaxZoom: getMinMaxZoom,
    getMinMaxZoomGDAL: getMinMaxZoomGDAL,
    getProjection: getProjection,
    projectionFromRaster: projectionFromRaster
};
